File: back/src/main/java/com/back/standard/recommend/byContent/RecommendByContent.java
package com.back.standard.recommend.byContent;

import com.back.standard.recommend.util.CosineSimilarityCalcer;
import com.back.standard.recommend.util.SimilarityCalcer;
import com.back.standard.recommend.util.Vector;

import java.util.*;

import static com.back.standard.recommend.util.SimilarityCalcer.Similar;

public class RecommendByContent {

    public record Content(
            Long id,
            String content
    ) { }

    private Map<Long, Vector> tfMap = new HashMap<>();
    private Map<Long, Integer> idfMap = new HashMap<>();

    private long totalCnt = 0;

    private SimilarityCalcer calcer = new CosineSimilarityCalcer();

    public RecommendByContent() {}
    public RecommendByContent(SimilarityCalcer calcer) { this.calcer = calcer; }

    private Vector getTFVector(String content) {

        totalCnt++;
        String[] words = content.split(" ");

        Map<Long, Integer> wordMap = new HashMap<>();

        for (String word : words) {
            long wordHash = word.hashCode();
            if (wordMap.containsKey(wordHash)) {
                wordMap.put(wordHash, wordMap.get(wordHash) + 1);
                continue;
            }
            wordMap.put(wordHash, 1);
            idfMap.put(wordHash, idfMap.getOrDefault(wordHash, 0) + 1);
        }

        Vector ret = new Vector();

        wordMap.forEach((key, value) ->
                ret.putValue(key, (float) value / wordMap.size()));

        return ret;
    }

    public void setData(List<Content> contents) {

        contents.forEach((content) -> {
            if (tfMap.containsKey(content.id())) return;
            tfMap.put(content.id(), getTFVector(content.content()));
        });

    }

    public List<Long> getRecommendList(Long id, int maxRecommends) {

        if (!tfMap.containsKey(id)) return List.of();

        Vector idf = new Vector();

        idfMap.forEach((key, value) -> {
            idf.putValue(key, Math.log((double) totalCnt / value));
        });

        Map<Long, Vector> tfIdf = new HashMap<>();

        tfMap.forEach((key, value) -> {
            tfIdf.put(key, Vector.hadamardProduct(value, idf));
        });

        List<Similar> similarList = calcer.getSimilarList(tfIdf.get(id), tfIdf);

        return similarList.stream().limit(maxRecommends).map(Similar::id).toList();
    }


}

File: back/src/main/kotlin/com/back/standard/recommend/byContent/RecommendByContent.kt
package com.back.standard.recommend.byContent

import com.back.standard.recommend.util.CosineSimilarityCalcer
import com.back.standard.recommend.util.SimilarityCalcer
import com.back.standard.recommend.util.Vector
import kotlin.math.ln

class RecommendByContent(
    private val calcer: SimilarityCalcer<Long, Long> = CosineSimilarityCalcer()
) {

    data class Content(
        val id: Long,
        val content: String
    )

    private val tfMap = mutableMapOf<Long, Vector<Long>>()
    private val idfMap = mutableMapOf<Long, Int>()

    private var totalCnt = 0L

    private fun getTFVector(content: String): Vector<Long> {
        totalCnt++
        val words = content.split(" ")

        val wordMap = mutableMapOf<Long, Int>()

        for (word in words) {
            val wordHash = word.hashCode().toLong()
            if (wordMap.containsKey(wordHash)) {
                wordMap[wordHash] = wordMap.getValue(wordHash) + 1
                continue
            }
            wordMap[wordHash] = 1
            idfMap[wordHash] = idfMap.getOrDefault(wordHash, 0) + 1
        }

        val ret = Vector<Long>()

        wordMap.forEach { (key, value) ->
            ret.putValue(key, value.toDouble() / wordMap.size)
        }

        return ret
    }

    fun setData(contents: List<Content>) {
        contents.forEach { content ->
            if (tfMap.containsKey(content.id)) return@forEach
            tfMap[content.id] = getTFVector(content.content)
        }
    }

    fun getRecommendList(id: Long, maxRecommends: Int): List<Long> {
        if (!tfMap.containsKey(id)) return emptyList()

        val idf = Vector<Long>()

        idfMap.forEach { (key, value) ->
            idf.putValue(key, ln(totalCnt.toDouble() / value))
        }

        val tfIdf = mutableMapOf<Long, Vector<Long>>()

        tfMap.forEach { (key, value) ->
            tfIdf[key] = Vector.hadamardProduct(value, idf)
        }

        val similarList = calcer.getSimilarList(tfIdf.getValue(id), tfIdf)

        return similarList.take(maxRecommends).map { it.id }
    }
}
